src/Schema/Directives/AllDirective.php

```diff
@@ -56,13 +58,48 @@ public function resolveField(FieldValue $fieldValue): FieldValue
                 $query = $this->getModelClass()::query();
             }
 
-            return $resolveInfo
+            $builder = $resolveInfo
                 ->argumentSet
                 ->enhanceBuilder(
                     $query,
                     $this->directiveArgValue('scopes', [])
-                )
-                ->get();
+                );
+
+            if (config('lighthouse.optimized_selects')) {
+                if (($builder instanceof QueryBuilder || $builder instanceof EloquentBuilder) && ! $this->directiveHasArgument('builder')) {
+                    $fieldSelection = array_keys($resolveInfo->getFieldSelection(1));
+
+                    $selectColumns = SelectHelper::getSelectColumns(
+                        $this->definitionNode,
+                        $fieldSelection,
+                        $this->getModelClass()
+                    );
+
+                    if (empty($selectColumns)) {
+                        return $builder->get();
+                    }
```

Comment:
Shouldn't this throw?

Reply:
In some cases,`$selectColumns` will be empty, and it shouldn’t throw exceptions because this is an expected behavior.

```graphql
type User {
    posts: [Post!]! #@hasMany
}

type Post {
    id: ID!
}

type Query {
    users: [User!]! @all
}
```
```graphql
{
    users {
        posts {
            id
        }
    }
}
```
Use the above schema as an example. The `posts: [Post!]!` will not be optimized due to empty `$selectColumns`. In the current implementation, the `@hasMany` directive is a must-have directive to let the optimized select work normally.

The optimized select is an improvement feature, even if this function is not working normally due to insufficient directives, it shouldn’t break original queries.

Reply:
In this example, the query would have to select the users `id` so that `posts` can be resolved without `@hasMany`, right?

In that case, I believe this check is insufficient in order to not break the query. If the client queries an additional field from `User`, `$selectColumns` would no longer be empty and result in `SELECT name FROM users`, thus breaking the resolution of `posts`.